### ISSUE TITLE
probe_dns_lookup_time_seconds wrong by orders of magnitude

### ISSUE THREAD

**Issue by PadaKwaak** [prometheus/blackbox_exporter#253]
### Host operating system:
I am using prom/blackbox-exporter docker image on a CentOS 7.3 host:
`uname -a`: `Linux mydockerhost.local 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
`cat /etc/centos-release`: `CentOS Linux release 7.3.1611 (Core)`
`docker --version`: `Docker version 17.07.0-ce, build 8784753`

### blackbox_exporter version:
I am using prom/blackbox-exporter:v0.10.0 docker image:
```
blackbox_exporter, version 0.10.0 (branch: HEAD, revision: 75681e3c4f051e24269a132e8ec19517118e1586)
  build user:       root@0334aae810a6
  build date:       20171009-14:25:49
  go version:       go1.9.1
```
The behaviour was the same with v0.7.0, v0.8.1, v0.9.0 & v0.9.1 docker images.

### What is the blackbox.yml module config.
```
modules:
  tcp_ipv4:
    prober: tcp
    tcp:
      preferred_ip_protocol: ip4
    timeout: 5s
```

### What is the prometheus.yml scrape config.
N/A, since I'm using `curl` from another host.

### What logging output did you get from adding `&debug=true` to the probe URL?
```
[root@chris ~]#  curl 'http://172.30.18.21:9115/probe?target=www.google.co.za:443&module=tcp_ipv4&debug=true'
Logs for the probe:
ts=2017-10-30T10:06:40.969596597Z caller=main.go:116 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Beginning probe" probe=tcp timeout_seconds=4.5
ts=2017-10-30T10:06:40.969686476Z caller=utils.go:43 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-30T10:06:41.188253005Z caller=utils.go:63 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Resolved target address" ip=216.58.223.35
ts=2017-10-30T10:06:41.188298125Z caller=tcp.go:55 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Dialing TCP without TLS"
ts=2017-10-30T10:06:41.436924007Z caller=tcp.go:99 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Successfully dialed"
ts=2017-10-30T10:06:41.437044619Z caller=main.go:127 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Probe succeeded" duration_seconds=0.467383866



Metrics that would have been returned:
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 2.03e-06
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.467383866
# HELP probe_failed_due_to_regex Indicates if probe failed due to regex
# TYPE probe_failed_due_to_regex gauge
probe_failed_due_to_regex 0
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 1



Module configuration:
prober: tcp
timeout: 5s
tcp:
  preferred_ip_protocol: ip4
```

### What did you do that produced an error?
Configured both the docker host and container to make use of a DNS server that is on another continent (North Virginia -> South Africa), so the roundtrip time is approximately 220ms.

### What did you expect to see?
`probe_dns_lookup_time_seconds` should have been like `2.18e-1`, since the DNS answer took 218ms. Even the debug logging confirmed this when you take the time difference between these 2 log entries:
```
ts=2017-10-30T10:06:40.969686476Z caller=utils.go:43 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-30T10:06:41.188253005Z caller=utils.go:63 module=tcp_ipv4 target=www.google.co.za:443 level=info msg="Resolved target address" ip=216.58.223.35
```

See the zipped tshark capture on the docker host of both the DNS lookup and TCP connect: [blackbox-exporter.zip](https://github.com/prometheus/blackbox_exporter/files/1426767/blackbox-exporter.zip)

I also expected blackbox-exporter's results to be similar to timing nslookup:
```
[root@mydockerhost root]# time nslookup www.google.co.za
Server:         172.30.0.36
Address:        172.30.0.36#53

Non-authoritative answer:
Name:   www.google.co.za
Address: 216.58.223.35


real    0m0.224s
user    0m0.003s
sys     0m0.003s
```

### What did you see instead?
`probe_dns_lookup_time_seconds 2.03e-06`, which is wrong by orders of magnitude

**Comment by SuperQ** [comment]
I can confirm there is something wrong here.  I used the same example.

```
ts=2017-10-30T15:13:10.615880638Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Beginning probe" probe=tcp timeout_seconds=4.5
ts=2017-10-30T15:13:10.621606058Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-30T15:13:11.202782445Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Resolved target address" ip=216.58.205.99
ts=2017-10-30T15:13:11.202834733Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Dialing TCP without TLS"
ts=2017-10-30T15:13:11.23884113Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Successfully dialed"
ts=2017-10-30T15:13:11.238983746Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Probe succeeded" duration_seconds=0.622989032
```

```
probe_dns_lookup_time_seconds 1.96e-07
probe_duration_seconds 0.622989032
```

The probe total duration is pretty close to the log line times:
* Log: 623.103ms
* Probe: 622.989ms

But the DNS lookup time doesn't match at all:
* Log: 581.176ms
* Probe: 0.000196ms

**Comment by SuperQ** [comment]
I tested a patch that changes the defer to an explicit update, this seems to return correct results:

```
ts=2017-10-30T15:28:13.05294489Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Beginning probe" probe=tcp timeout_seconds=9.5
ts=2017-10-30T15:28:13.053098713Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Resolving target address" preferred_ip_protocol=ip4
ts=2017-10-30T15:28:13.71350531Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Resolved target address" ip=216.58.205.99
ts=2017-10-30T15:28:13.713629436Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Dialing TCP without TLS"
ts=2017-10-30T15:28:13.748827495Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Successfully dialed"
ts=2017-10-30T15:28:13.749069157Z caller=main.go:174 module=tcp_ipv4 target=www.google.co.za:443 level=debug msg="Probe succeeded" duration_seconds=0.695987381
```

```
probe_dns_lookup_time_seconds 0.660433915
probe_duration_seconds 0.695987381
```

New DNS lookup time:
* Log: 660.407ms
* Probe: 660.434ms

There's still a few microseconds of deviation here, but it's correct because in my test I put the metric change after the log line.
